cylc/flow/main_loop/__init__.py

```diff
@@ -0,0 +1,373 @@
+
+
+def _debounce(interval, timings):
+    """Rate limiter, returns True if the interval has elapsed.
+
+    Arguments:
+        interval (float):
+            Time interval in seconds as a float-type object.
+        timings (list):
+            List-list object of the timings of previous runs in the form
+            ``(completion_wallclock_time, run_duration)``.
+            Wallclock times are unix epoch times in seconds.
+
+    Examples:
+        >>> from time import time
+
+        No previous run (should always return True):
+        >>> _debounce(1., [(0, 0)])
+        True
+
+        Interval not yet elapsed since previous run:
+        >>> _debounce(1., [(time(), 0)])
+        False
+
+        Interval has elapsed since previous run:
+        >>> _debounce(1., [(time() - 2, 0)])
+        True
+
+    """
+    if not interval:
+        return True
+    try:
+        last_run_at = timings[-1][0]
+    except IndexError:
+        last_run_at = 0
+    if (time() - last_run_at) > interval:
+        return True
+    return False
+
+
+def startup(fcn):
+    """Decorates a coroutine which is run at suite startup.
+
+    The decorated coroutine should have the signature:
+
+        ``async coroutine(scheduler, plugin_state) -> None``
+
+    Exceptions:
+
+        * Regular Exceptions are caught and logged.
+        * Exceptions which subclass CylcError are re-raised as
+          MainLoopPluginException
+
+    """
+    fcn.main_loop = CoroTypes.StartUp
+    return fcn
+
+
+def shutdown(fcn):
+    """Decorates a coroutine which is run at suite shutdown.
+
+    Note shutdown refers to "clean" shutdown as opposed to suite abort.
+
+    The decorated coroutine should have the signature:
+
+        ``async coroutine(scheduler, plugin_state) -> None``
+
+    Exceptions:
+
+        * Regular Exceptions are caught and logged.
+        * Exceptions which subclass CylcError are re-raised as
+          MainLoopPluginException
+
+    """
+    fcn.main_loop = CoroTypes.ShutDown
+    return fcn
+
+
+def periodic(fcn):
+    """Decorates a coroutine which is run at a set interval.
+
+    The decorated coroutine should have the signature:
+
+        ``async coroutine(scheduler, plugin_state) -> None``
+
+    Exceptions:
+
+        * Regular Exceptions are caught and logged.
+        * Exceptions which subclass CylcError are re-raised as
+          MainLoopPluginException
+
+    Configuration:
+
+        * The interval of execution can be altered using the
+          `cylc[main loop][plugin name]interval` setting.
+
+    """
+    fcn.main_loop = CoroTypes.Periodic
+    return fcn
+
+
+class CoroTypes:
+    """Different types of coroutine which can be used with the main loop."""
+
+    StartUp = startup
+    ShutDown = shutdown
+    Periodic = periodic
+
+
+def load(config, additional_plugins=None):
+    additional_plugins = additional_plugins or []
+    entry_points = pkg_resources.get_entry_map(
+        'cylc-flow'
+    ).get('main_loop', {})
```

Comment:
Couldn't use the `my-plugin` from the documentation. Debugging, the `entry_points` here never had my `my-plugin`, so that's never loaded correctly for me.

Is there anything I am missing @oliver-sanders ? I thought we would need something like this https://packaging.python.org/guides/creating-and-discovering-plugins/ (see part about namespaces)

Reply:
So, I think if you install this in the Cylc Flow `setup.cfg` file then it'll work, however, we haven't yet implemented the mechanism required for external projects to add entrypoints to Cylc Flow (I don't think).

The only experience I have with this is with Pytest, which uses the entrypoint `pytest11`.

Reply:
Let's move that for later than :+1: